crypto/cmac/cm_meth.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include "internal/cryptlib.h"
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include <openssl/evp.h>
+#include <openssl/cmac.h>
+#include "internal/evp_int.h"
+
+/* local CMAC pkey structure */
+
+/* typedef EVP_MAC_IMPL */
+struct evp_mac_impl_st {
+    /* tmpcipher and tmpengine are set to NULL after a CMAC_Init call */
+    const EVP_CIPHER *tmpcipher; /* cached CMAC cipher */
+    const ENGINE *tmpengine;     /* cached CMAC cipher engine */
+    CMAC_CTX *ctx;
+};
+
+static EVP_MAC_IMPL *cmac_new(void)
+{
+    EVP_MAC_IMPL *cctx;
+
+    if ((cctx = OPENSSL_zalloc(sizeof(*cctx))) == NULL
+        || (cctx->ctx = CMAC_CTX_new()) == NULL) {
+        OPENSSL_free(cctx);
+        cctx = NULL;
+    }
+
+    return cctx;
+}
+
+static void cmac_free(EVP_MAC_IMPL *cctx)
+{
+    if (cctx != NULL) {
+        CMAC_CTX_free(cctx->ctx);
+        OPENSSL_free(cctx);
+    }
+}
+
+static int cmac_copy(EVP_MAC_IMPL *cdst, EVP_MAC_IMPL *csrc)
+{
+    if (!CMAC_CTX_copy(cdst->ctx, csrc->ctx))
+        return 0;
+
+    cdst->tmpengine = csrc->tmpengine;
+    cdst->tmpcipher = csrc->tmpcipher;
+    return 1;
+}
+
+static size_t cmac_size(EVP_MAC_IMPL *cctx)
+{
+    return EVP_CIPHER_CTX_block_size(CMAC_CTX_get0_cipher_ctx(cctx->ctx));
+}
+
+static int cmac_init(EVP_MAC_IMPL *cctx)
+{
+    int rv = CMAC_Init(cctx->ctx, NULL, 0, cctx->tmpcipher,
+                       (ENGINE *)cctx->tmpengine);
+    cctx->tmpcipher = NULL;
+    cctx->tmpengine = NULL;
+
+    if (!rv)
```

Comment:
Would `return rv;` work here?  `CMAC_Init` returns 0 or 1.
Or even `return rv ? 1 : 0;` ?

Similarly for the next couple of functions.

Reply:
Sure.

Reply:
The other macs (HMAC, ...) had similar.